File: leetcode/0009/python3/main.py
# coding=utf-8

class Solution:
    def isPalindrome(self, x: 'int') -> 'bool':
        if x < 0:
            return False

        xs = str(x)
        rxs = xs[::-1]
        return xs == rxs

if __name__ == "__main__":
    s = Solution()

    assert(s.isPalindrome(121) == True);
    assert(s.isPalindrome(-121) == False);
    assert(s.isPalindrome(10) == False);
    assert(s.isPalindrome(0) == True);
    assert(s.isPalindrome(1) == True);
    assert(s.isPalindrome(11) == True);

File: leetcode/0009/cpp/main.cpp
#include <cassert>
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0)
        {
            return false;
        }

        if (x < 10)
        {
            return true;
        }

        vector<int> nums;

        while (x != 0)
        {
            int r = x % 10;
            nums.push_back(r);
            x /= 10;
        }

        uint32_t first = 0;
        uint32_t last = static_cast<uint32_t>(nums.size() - 1);
        while (first < last)
        {
            if (nums[first] != nums[last])
            {
                return false;
            }

            ++first;
            --last;
        }

        return true;
    }
};

int main()
{
    Solution s;

    assert(s.isPalindrome(121) == true);
    assert(s.isPalindrome(-121) == false);
    assert(s.isPalindrome(10) == false);
    assert(s.isPalindrome(0) == true);
    assert(s.isPalindrome(1) == true);
    assert(s.isPalindrome(11) == true);

    return 0;
}
